pkg/agent/openflow/network_policy_test.go

```diff
@@ -282,91 +292,286 @@ func TestInstallPolicyRuleFlows(t *testing.T) {
 }
 
 func TestBatchInstallPolicyRuleFlows(t *testing.T) {
-	ctrl := gomock.NewController(t)
-	defer ctrl.Finish()
-
-	c = prepareClient(ctrl)
-	c.nodeConfig = &config.NodeConfig{PodIPv4CIDR: podIPv4CIDR, PodIPv6CIDR: nil}
-	c.ipProtocols = []binding.Protocol{binding.ProtocolIP}
-	defaultAction := crdv1alpha1.RuleActionAllow
-	priorityRule2 := uint16(10000)
-
-	ruleID1 := uint32(10)
-	rule1 := &types.PolicyRule{
-		Direction: v1beta2.DirectionOut,
-		From:      parseAddresses([]string{"192.168.1.40", "192.168.1.50"}),
-		Action:    &defaultAction,
-		To:        parseAddresses([]string{"0.0.0.0/0"}),
-		FlowID:    ruleID1,
-		TableID:   EgressRuleTable,
-		PolicyRef: &v1beta2.NetworkPolicyReference{
-			Type:      v1beta2.K8sNetworkPolicy,
-			Namespace: "ns1",
-			Name:      "np1",
-			UID:       "id1",
+	tests := []struct {
+		name            string
+		rules           []*types.PolicyRule
+		expectedFlowsFn func(c *client) []binding.Flow
+	}{
+		{
+			name: "multiple K8s NetworkPolicy rules",
+			rules: []*types.PolicyRule{
+				{
+					Direction: v1beta2.DirectionOut,
+					From:      parseAddresses([]string{"192.168.1.40", "192.168.1.50"}),
+					To:        parseAddresses([]string{"0.0.0.0/0"}),
+					FlowID:    uint32(10),
+					TableID:   EgressRuleTable,
+					PolicyRef: &v1beta2.NetworkPolicyReference{
+						Type:      v1beta2.K8sNetworkPolicy,
+						Namespace: "ns1",
+						Name:      "np1",
+						UID:       "id1",
+					},
+				},
+				{
+					Direction: v1beta2.DirectionOut,
+					// conjunctive match flow for nw_src=192.168.1.40 should commit to conjunction 10 and 11.
+					// conjunctive match flow for nw_src=192.168.1.45 should commit to conjunction 11 only.
+					From:    parseAddresses([]string{"192.168.1.40", "192.168.1.51"}),
+					To:      parseAddresses([]string{"0.0.0.0/0"}),
+					Service: []v1beta2.Service{{Protocol: &protocolTCP, Port: &port8080}},
+					FlowID:  uint32(11),
+					TableID: EgressRuleTable,
+					PolicyRef: &v1beta2.NetworkPolicyReference{
+						Type:      v1beta2.K8sNetworkPolicy,
+						Namespace: "ns1",
+						Name:      "np2",
+						UID:       "id2",
+					},
+				},
+			},
+			expectedFlowsFn: func(c *client) []binding.Flow {
```

Comment:
do you think it would be possible to generate these expectations in a more human-readable way? Right now there is a lot of noise because of the cookie stuff, etc.

in other words would it be possible to have something like this:
```
type ExpectedFlowsBuilder struct {...}

fb := &ExpectedFlowsBuilder{}
  .AddConjunction(10).AddConjunction(11)
  .AddConjunctiveMatch(<match info>, ConjunctionAction(10, 1, 2), ConjunctionAction(11, 1, 3))
  .AddConjunctiveMatch(<match info>, ConjunctionAction(11, 1, 3))
...
fb.Flows() // return the list of Flows
```

It would take care of generating metric flows and all the "noise".

Reply:
The conjunction rule may have different in priority and table as well. If too customized for NetworkPolicy flow, it seems we will need such builders when we add verficiation for other flows, like service, pod. But if it's generic, it would be quite similar to the current FlowBuilder, and maybe we should just make FlowBuilder more simpler to use, removing some redundant part.
Maybe I missed your point, I can follow up after the fix is merged, and feel free to update the code directly if you have some mature idea.

Reply:
> Maybe I missed your point, I can follow up after the fix is merged, and feel free to update the code directly if you have some mature idea.

No, it seems that we are on the same page. I feared it may not be as simple as I hoped :) Definitely doesn't need to be part of this PR. It may be indeed be an opportunity to simplify the FlowBuilder, although I feel like some customization for testing would always be needed. Ideally the expectations I define would not need to define the metrics flow separately for example.